### PR TITLE
Simplify with recurrent mutations, etc.

### PR DESCRIPTION
**Pull request by jeromekelleher**
Fix up the mess in simplify caused by handling of recurrent mutations, etc.

This adds a ``parent`` column to the mutation table, which is the ID of a given mutation's parent mutation. Hopefully this will make clearing up this mess easier, and clear the way for nice general methods for statistics.

### CODE REVIEW

**Comment by jeromekelleher** [comment]
OK, I _think_ this is nearly done. It would like to say it's been fun, but it hasn't; this is a _hard_ and subtle problem!

There are two outstanding issues I could use your input on @petrelharp:

1. The ``filter_invariant_sites`` flag is now misleading. What I have implemented is really ``filter_zero_mutation_sites``, because actually getting all invariant sites in general is hard; we can't be sure of this unless we actually create the tree and look at the sample genotypes, essentially. I think we should rename this flag to ``filter_zero_mutation_sites``
2. More fundamentally, it appears that we have to make a choice between either having the samples provided to simplify be mapped to 0, ..., n - 1 in the output tree sequence **or** we enforce that edges are listed in increasing order of parent ID in the output edgesets. I think we have to let go of the parent ordering within edgesets, as it's just too much trouble.

What do you think?

**Comment by jeromekelleher** [comment]
I've updated this to disable the check for sorted parents, and it all seems to work OK. If you agree with this @petrelharp I'll push through the change more rigorously.

Also note I've cherry picked in the WF mutations from #248.

**Comment by petrelharp** [comment]
(1) `filter_zero_mutation_sites` - sure, sounds good, although `filter_invariant_sites` is not wrong, since all sites filtered *are* invariant sites.  A shorter name would be `filter_unmutated_sites`.

(2) I'm not bothered by a lack of sorted parents.  Still sorted by parent time, though?

**Comment by jeromekelleher** [comment]
> (1) filter_zero_mutation_sites - sure, sounds good, although filter_invariant_sites is not wrong, since all sites filtered are invariant sites. A shorter name would be filter_unmutated_sites.

Well ``filter_invariant_sites`` would imply to me that it should filter all invariant sites, so I think we should avoid that. We can add further flags in the future to filter out other types of sites,  so I think it's good to be precise here (even at the expense of it being a bit wordy).

> (2) I'm not bothered by a lack of sorted parents. Still sorted by parent time, though?

Yes, definitely. And parents should be grouped together, otherwise simplify won't work.

**Comment by jeromekelleher** [comment]
OK, I've updated those points @petrelharp and I think we're ready to merge.

File: lib/table.c [hunk 1]
```diff
@@ -790,6 +796,12 @@ mutation_table_append_columns(mutation_table_t *self, size_t num_rows, site_id_t
             num_rows * sizeof(node_id_t));
     memcpy(self->derived_state + self->total_derived_state_length, derived_state,
             total_derived_state_length * sizeof(char));
+    if (parent == NULL) {
+        /* If parent is NULL, set all parents to the null mutation */
+        memset(self->parent + self->num_rows, 0xff, num_rows * sizeof(mutation_id_t));
```
Comment (petrelharp) [inline lib/table.c]: Why is this the null mutation?

Comment (jeromekelleher) [reply lib/table.c]: Why do we set the parent to the null mutation by default, or why is 0xff the null mutation?

Comment (petrelharp) [reply lib/table.c]: Sorry, the second -- why 0xff?

Comment (jeromekelleher) [reply lib/table.c]: The null mutation/node is -1, which is all 1s at the bit level. 0xf is hex for all 1s. Memset takes a byte as it's argument, so you have to specify 0xff to fill a buffer with 1 bits. It's not that obvious, and is a departure from using NULL_MUTATION where possible but it is convenient and fast.

**Comment by petrelharp** [comment]
thanks!
